### DISCUSSION TITLE
[apex] Could NoLogicInTrigger Rule recognize Platform Event triggers?

### DISCUSSION THREAD

**Discussion by DeborahOrth** [pmd/pmd#3129] [General]
I don't seem to be able to suppress the NoLogicInTrigger rule for Platform Event type triggers.  I don't see the community using a trigger handler class for PE, instead the code and Salesforce examples I am seeing are all including the trigger logic directly in the trigger.  So, could we get this rule to recognize **__e** triggers and not show the rule violation?

**Comment by adangel** [comment]
Thanks for the question. As I'm not familiar with Apex, could you give a code sample?

**Reply by DeborahOrth** [reply]
Thanks for getting back to me.  

On further investigation into this issue, I think I am stuck just having the PMD rule fire for this no matter what.  When trying to use **@SuppressWarnings('PMD.AvoidLogicInTrigger')** before a trigger definition, (event or object), PMD does bypass the rule but the code won't compile on Salesforce, so either we live with the warning or disable the rule entirely.

Here are some details about Apex and what I am talking about:

Apex is a the proprietary server-side language used by Salesforce.com.  It is based off Java but includes a native database layer, testing framework, etc.  The SF database is an object oriented DB and triggers are also written in Apex.

A typical trigger module might look something like this:
```
trigger MyCustomObjectTrigger on MyCustomObject__c (before insert, after insert, before update, after update){
   if (Trigger.isBefore){
      if (Trigger.isInsert){
         MyCustomObjectTriggerHandler.beforeInsert(Trigger.new);
      } else if (Trigger.isUpdate){
         MyCustomObjectTriggerHandler.beforeUpdate(Trigger.old, Trigger.newMap);
      }
...
}
```
There is also an Event Bus that uses triggers inside of Salesforce to subscribe to events, the code looks very similar to a regular object trigger, except it can only handle the **after insert** event, here is an example of the declaration:
```
trigger MyEventTrigger on MyEvent__e (after insert){
   ...
}
```

Apex coding conventions say that no Object trigger should have logic in it, a handler class should be used.  But in the Event type triggers, I am not seeing that coding convention and I would like to suppress the warning but... back to compiler issue :frowning:. I guess I could just refactor code into handler anyhow, even if no one else seems to do this.

**Comment by rsoesemann** [comment]
Hi @DeborahOrth, you cannot take the examples of the community not even the code examples of Salesforce or Trailhead as best practices. It took Salesforce years to officially document that SObject triggers should be logicless and they might do that for event triggers as well. In some years...

Triggers of what every type should have no logic and delegate everything into Apex classes. Most trigger handler frameworks perfectly work for Custom and even Standard Platform events and have exactely one line like this:

```
trigger MyCustomEvent on MyCustomEvent__e (after insert) {
    new MyCustomEventHandler().handle();
}
```

Why can I answer this question quite confident? I was creating/porting all initial PMD rules for Apex and work as a Salesforce architect for nearly 10 years now.

**Reply by DeborahOrth** [reply]
Thanks @rsoesemann, I know the SF docs are not the best, I just haven't seen any code example like the one you show, good to know that this is an unpublished best practice, I will certainly use it for my code.  

I too have been working in SF for 10 years and still find conflicting docs on what the best practices are!  I do love PMD and appreciate all the work you do to promote it within the SF community!
